### PR TITLE
Status changes 124

### PR DESCRIPTION
**Pull request by evilnick**
This is part one of the status changes for 1.24 of Juju. It adds:

 - The status-set and status-get hook tools, and explanations
 - A reference table detailing the different status values (including ones set by Juju)

N.B., the table is ugly. Current site CSS overrides sensible table styles and forces it, for example, to be aligned centre :( 
I think this will be fixed in a soon-to-be-released version of jujucharms.com

### CODE REVIEW

File: src/en/authors-hook-environment.md [hunk 1]
```diff
@@ -479,3 +483,89 @@ relation-ids reverseproxy
...
+A unit in the `active` state with should not generally expect anyone to look at
+its status message, and often it is better not to set one at all. In the event
+of a degradation of service, this is a good place to surface an explanation for
+the degradation (load, hardware failure or other issue).
+
+A unit in `error` state will have a message that is set by Juju and not the
+charm because the error state represents a crash in a charm hook - an unmanaged
+and uninterpretable situation. Juju will set the message to be a reflection of
+the hook which crashed. For example “Crashed installing the software” for an
+install hook crash, or “Crash establishing db link to postgres” for a crash in a
```
Comment (petermatulis) [inline src/en/authors-hook-environment.md]: PostgreSQL?
postgresql?

File: src/en/authors-hook-environment.md [hunk 2]
```diff
@@ -479,3 +483,89 @@ relation-ids reverseproxy
...
 `--format yaml`, and you may consider it smarter to use those for clarity's 
 sake than to depend on the default `smart` format.
+
+
+### status-set
+
+Introduced in version 1.24 of Juju, a new status mechanism allows Juju and its 
+charms to more accurately reflect their current status. This places the 
+responsibility on the charm to know its status, and set it accordingly using
+the `status-set` hook tool.
```
Comment (petermatulis) [inline src/en/authors-hook-environment.md]: It's unclear whether this more detailed status will be included in the output of a standard 'juju status' on the part of a user.

File: src/en/authors-hook-environment.md (hunk 1 above)
Comment (marcoceppi) [inline src/en/authors-hook-environment.md]: The charm is postgresql the service is PostgreSQL

File: src/en/authors-hook-environment.md [hunk 3]
```diff
@@ -479,3 +483,89 @@ relation-ids reverseproxy
...
+
+The second argument is a user-facing message, which will be displayed to any
+users viewing the status, and will also be visible in the status history. This
+can contain any useful information - 
+
+This status message provides valuable feedback to the user about what is
+happening. Changes in the status message are not broadcast to peers and
+counterpart units - they are for the benefit of humans only, so tools
+representing Juju services should ask occasionally and be told the current
+status message.
```
Comment (petermatulis) [inline src/en/authors-hook-environment.md]: > ...so tools representing Juju services should ask occasionally and be told the current status message.

I don't know what this phrase means.

File: src/en/authors-hook-environment.md [hunk 4]
```diff
@@ -479,3 +483,89 @@ relation-ids reverseproxy
...
+### status-get
+
+The status-get hook tool allows a charm to query what is recorded in Juju as the
+current workload status. Without arguments, it just prints the workload status
+value eg maintenance. With `--include-data` specified, it prints YAML which 
+contains the status value plus any data associated with the status.
+
+Examples:
+
+```nohighlight
```
Comment (petermatulis) [inline src/en/authors-hook-environment.md]: Consistency. Are we using 'no-highlight' or 'nohighlight'?

File: src/en/reference-status.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
...oviding services, but is actively doing stuff in preparation for providing those services. This is a “spinning” state, not an error state. It reflects activity on the unit itself, not on peers or related units. The Juju agent will set the unit to maintenance when it begins the process of installing the charm - downloading it, unpacking it and calling the install hook.<br>The unit may transition to the maintenance state at any time, in any hook. The maintenance state implies that the unit has no reason to think that it will not provide services soon. For example, the software might be formatting block devices, or replicating data from peer units, or handling an action which for whatever reason means that it is not actively providing service, but it is not aware of anything on its own config or in related unit config that will prevent it from bringing up its services.<br>Counterparts to this unit should expect that service calls to this unit will fail during the period it is in the maintenance state, but unless they have an urgent need to move to another counterpart, they should just wait for this unit to finish churning.<br> Being in maintenance does not require human intervention.|
```
Comment (petermatulis) [inline src/en/reference-status.md]: PostgreSQL

File: src/en/reference-status.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+
+# Unit status
+
+As of version 1.24 of Juju, a unit can report its status as one of the
+following:
+
+| Status       | colour in GUI* | Meaning                                      |
+| :----------- | :------------- | :------------------------------------------- |
+| blocked      | blue           | This unit needs manual intervention because of a problem associated directly with the unit itself. This may be due to bad or inconsistent configuration of the service, disk space being unavailable for the unit, or missing relations which are essential to service operations. The message should tell a human administrator what is required to unblock the service.<br> For example: “Please provide an object store. Please configure the minimum speed required by users of this service. Please link this service to a database like MySQL or Postgres.”|
```
Comment (petermatulis) [inline src/en/reference-status.md]: PostgreSQL

File: src/en/reference-status.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+|waiting      |yellow         | The unit is unable to progress to an active state because a service to which it is related is not running. That service might be in error, blocked, waiting or maintenance. This unit expects it will be able to proceed towards active as soon as the things it is concerned about are resolved there - it needs no human attention (otherwise its status would be blocked).|
+|active |green| This unit believes it is correctly offering all the services it is primarily installed to provide. A service can be active even if additional relations are being configured and setup. For example, a ceph osd service that is fully connected up to its controllers is active, even if you then relate it to nagios; setting up the monitoring subordinate does not impinge on the ceph osd functionality which is its primary purpose for existing.|
```
Comment (petermatulis) [inline src/en/reference-status.md]: Ceph OSD (2 places), or just "OSD" (remove "ceph") for the second instance
Nagios

"fully connected up to" - odd. Maybe "fully joined with", or just "joined with"

"[Ceph OSD] controllers" - I'm not sure what this refers to. Once set up, an OSD services Ceph client requests. Is this referring to bringing up an OSD? If so, it is probably referring to its "Monitors".

Personally, "setup" is a noun and "set up" is a verb. Same for "login" and "log in".

File: src/en/authors-hook-environment.md (hunk 2 above)
Comment (evilnick) [inline src/en/authors-hook-environment.md]: the more detailed status will be exposed to anything that checks status, so yes, it does include the command-line tool. I don't think that really needs to be spelled out here, and it is more obvious from the following paragraphs anyhow

Comment (petermatulis) [inline src/en/authors-hook-environment.md]: This is an authors' page (which I didn't appreciate when I wrote the comment) so I agree.

**Comment by petermatulis** [comment]
:+1:

**Comment by evilnick** [comment]
\o/

On 23 June 2015 at 15:42, Peter Matulis <notifications@github.com> wrote:

> Merged #507 <https://github.com/juju/docs/pull/507>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juju/docs/pull/507#event-337898161>.
>



-- 
Nick Veitch
nick.veitch@canonical.com

**Pull request merged**